### PR TITLE
SonarQube integration fix

### PR DESCRIPTION
**Pull request by dgeorgievski**
SonarQube integration fails due to expectation by the plugin to read SQ task info from the build logs. Instead this info is available in the sonar build report `${WORKSPACE}/build/sonar/report-task.txt`

A workaround in the form of a documentation for InflluxDB v2 integration that changes the semantics for user authentication. 
See the  `doc/influxdb-v2-integration.md` for more details.

I've also added a new field for JUnit metric counter called `test_count` that makes aggregations in reports much easier. 

Related to https://issues.jenkins.io/browse/JENKINS-67051

- [x ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x ] Please describe what you did
- [x ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

Tests has been performed with the following pipeline on Jenkins  v2.289 with SonarQUbe Version 8.9.3 (build 48735), and InfluxDB v2.0

pipeline
```
pipeline {
    agent any    

    stages {
        stage('Clone sources') {
            steps {
                git url: 'https://github.com/dgeorgievski/sonarqube-jacoco-code-coverage.git'
            }
        }        
        
        stage('Build') {
            steps {
                sh './gradlew clean test build'
                junit 'build/test-results/test/*.xml'
                step( [ $class: 'JacocoPublisher' ] )
            }
        }
        
        stage('SonarQube analysis') {
            steps {
                withSonarQubeEnv('SonarQube') {
                    sh "./gradlew sonarqube -PsonarToken=**********************"
                }
            }
        }
        
        stage("Quality gate") {
            steps {
                waitForQualityGate abortPipeline: true
            }
        }
        
        stage("InfluxDB v2 publisher") {
            environment {
                LOG_JUNIT_RESULTS="true"
                INFLUXDB_PLUGIN_CUSTOM_PROJECT_NAME = "foo"
            }
            
            steps {
                withSonarQubeEnv('SonarQube') {
                    influxDbPublisher(selectedTarget: 'influxdb_v2',)
                }
            }
        }
        
    }
}
```

The following is a fragment of the Console output relevant to SonarQube and InfluxDB
```
[Pipeline] }
[Pipeline] // withSonarQubeEnv
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (Quality gate)
[Pipeline] waitForQualityGate
Checking status of SonarQube task 'AX06VyrQXKUx1dTNinuS' on server 'SonarQube'
SonarQube task 'AX06VyrQXKUx1dTNinuS' status is 'IN_PROGRESS'
SonarQube task 'AX06VyrQXKUx1dTNinuS' status is 'SUCCESS'
SonarQube task 'AX06VyrQXKUx1dTNinuS' completed. Quality gate is 'OK'
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (InfluxDB v2 publisher)
[Pipeline] withEnv
[Pipeline] {
[Pipeline] withSonarQubeEnv
Injecting SonarQube environment variables using the configuration: SonarQube
[Pipeline] {
[Pipeline] influxDbPublisher
[InfluxDB Plugin] Collecting data...
[InfluxDB Plugin] JaCoCo data found. Writing to InfluxDB...
[InfluxDB Plugin] JUnit data found. Writing to InfluxDB...
[InfluxDB Plugin] SonarQube hasReport
[InfluxDB Plugin] SonarQube data found. Writing to InfluxDB...
[InfluxDB Plugin] INFO: Using SonarQube host URL: http://localhost:9000
[InfluxDB Plugin] INFO: Using SonarQube auth token found in environment variable SONAR_AUTH_TOKEN
[InfluxDB Plugin] INFO: SonarQube task AX06VyrQXKUx1dTNinuS status is SUCCESS
[InfluxDB Plugin] Publishing data to target 'influxdb_v2' (url='http://10.73.71.63:8186', database='quality-metrics')
[InfluxDB Plugin] Completed.
[Pipeline] }
[Pipeline] // withSonarQubeEnv
[Pipeline] }
[Pipeline] // withEnv
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Finished: SUCCESS
```

### CODE REVIEW

File: doc/SonarQube-Integration.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  }
+}
+```
+
+## Development
+### Build
+mvn clean package hpi:hpi
+
+### Deploy
+Use the Advanced tab in `Manage Plugins` page to upload `target/influxdb.hpi` file. Restart Jenkins.
```
Comment (asimell) [inline doc/SonarQube-Integration.md]: This has nothing to do with SonarQube integration, so it should not be part of this document. 99% of users are interested in using the plugin, not developing it. The only relevant part of this section is
```
withSonarQubeEnv('SonarQube') {
    influxDbPublisher(selectedTarget: 'influxdb_v2',)
}
```
which specifies that you need to run `InfluxDbPublisher` in the same `withSonarQubeEnv` block as you ran SonarQube analysis.

File: doc/influxdb-v2-integration.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
+# InfluxDB v2 Integration
+
+The current plugin version uses InfluxDB v1.x authentication semantics (user name and password, and a database), while v2 is using APi token, bucket, and organization to regulate the acceess to the database.
+
+The InfluxDB Target configuration options deceptively indicate InfluDB v2 options, but it looks like the plugin still expects to connect to InfluxDB v1.x. See this ticket for more details https://issues.jenkins.io/browse/JENKINS-65830
```
Comment (asimell) [inline doc/influxdb-v2-integration.md]: How is it deceptive? InfluxDB 2 allows authentication with username and password, even though token based authentication is the standard. Token based authentication is also on the roadmap, so I think an extensive documentation for a workaround is a massive overkill.

This plugin is also aimed to send data into a readily configured InfluxDB database. Instructions on how to set up the database is not really a job for this plugin. [Influxdata](https://docs.influxdata.com/influxdb/v2.0/install/) has pretty good instructions on how to properly configure InfluxDB.

File: src/main/java/jenkinsci/plugins/influxdb/generators/JUnitPointGenerator.java [hunk 3]
```diff
@@ -63,6 +64,7 @@ public boolean hasReport() {
                     .addField(JUNIT_TEST_STATUS, caseResult.getStatus().toString())
                     .addField(JUNIT_TEST_STATUS_ORDINAL, caseResult.getStatus().ordinal())
                     .addField(JUNIT_DURATION, caseResult.getDuration())
+                    .addField(JUNIT_COUNT, 1)
```
Comment (asimell) [inline src/main/java/jenkinsci/plugins/influxdb/generators/JUnitPointGenerator.java]: Make sure you also add relevant documentation to [available metrics documentation](https://github.com/jenkinsci/influxdb-plugin/blob/development/doc/available_metrics.md).

File: src/test/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGeneratorTest.java [hunk 4]
```diff
@@ -39,42 +40,46 @@ public void before() {
...
-        assertEquals(gen.getSonarProjectName(url), name);
-    }
+    // @Test
+    // public void getSonarProjectNameFromNewSonarQube() throws Exception {
+    //     EnvVars envVars = new EnvVars();
+    //     String name = "org.namespace:feature%2Fmy-sub-project";
+    //     String url = sonarUrl + "/dashboard?id=" + name;
+    //     SonarQubePointGenerator gen = new SonarQubePointGenerator(build, listener, measurementRenderer, currTime, StringUtils.EMPTY, CUSTOM_PREFIX, envVars);
+    //     assertEquals(gen.getSonarProjectName(url), name);
+    // }
```
Comment (asimell) [inline src/test/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGeneratorTest.java]: Why have you commented out a unit test?

File: src/test/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGeneratorTest.java [hunk 5]
```diff
@@ -39,42 +40,46 @@ public void before() {
...
-        assertEquals(gen.getSonarProjectName(url), name);
-    }
+    // @Test
+    // public void getSonarProjectName() throws Exception {
+    //     EnvVars envVars = new EnvVars();
+    //     String name = "org.namespace:feature%2Fmy-sub-project";
+    //     String url = sonarUrl + "/dashboard/index/" + name;
+    //     SonarQubePointGenerator gen = new SonarQubePointGenerator(build, listener, measurementRenderer, currTime, StringUtils.EMPTY, CUSTOM_PREFIX, envVars);
+    //     assertEquals(gen.getSonarProjectName(url), name);
+    // }
```
Comment (asimell) [inline src/test/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGeneratorTest.java]: Why have you commented out a unit test?

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java [hunk 6]
```diff
@@ -235,63 +229,75 @@ protected String getResult(String url) throws IOException {
                 .url(url)
                 .header("Accept", "application/json");
 
+        // String logMessage = "[InfluxDB Plugin]: No SonarQube auth token found in environment variable SONAR_AUTH_TOKEN. Depending on access rights, this might result in a HTTP/401.";
```
Comment (asimell) [inline src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: Please remove unnecessary commented command

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java [hunk 7]
```diff
@@ -235,63 +229,75 @@ protected String getResult(String url) throws IOException {
...
+
             if (response.code() != 200) {
                 throw new RuntimeException("Failed : HTTP error code : " + response.code() + " from URL : " + url);
             }
             ResponseBody body = response.body();
             if (body == null) {
                 throw new NullPointerException("Failed : null body from URL : " + url);
             }
-
+
```
Comment (asimell) [inline src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: Remove trailing whitespace

**Review by asimell — CHANGES_REQUESTED** [review]
Thanks for the PR, please fix/answer the added comments and also the SpotBugs issue.

File: src/main/java/jenkinsci/plugins/influxdb/generators/JUnitPointGenerator.java (hunk 3 above)
Comment (dgeorgievski) [reply src/main/java/jenkinsci/plugins/influxdb/generators/JUnitPointGenerator.java]: Done.

File: doc/influxdb-v2-integration.md (hunk 2 above)
Comment (dgeorgievski) [reply doc/influxdb-v2-integration.md]: I was probably channeling my initial frustration with using the plugin – I thought it would be easy to configure and use it with InfluxDB v2, but I just couldn’t make the authentication work. The target configuration was “deceptively” simple 😊. 

Now, that I’ve reviewed the code I wouldn’t use the same wording but I’d focus on improving the documentation to define explicitly the supported type of authentication. It would help newbies like me to adopt influxdb_plugin faster. 

I did a brief review of the InfluxDB [influxdb-client-java](https://github.com/influxdata/influxdb-client-java/blob/2d5ca33ef678a8e6cc14bca6556ab2723fb54ce5/client/README.md) and I would say it relies strictly on tokens to authenticate clients.  

The version used by influxdb_client is old - 1.15.0. Although my initial focus was on SonarQube integration I’d be more than happy to update the client and implement the token-based authentication.  

In any case, the documentation was re-phrased and I tried to provide more context on the challenge and solution.

File: src/test/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGeneratorTest.java (hunk 4 above)
Comment (dgeorgievski) [reply src/test/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGeneratorTest.java]: Removed. Initially I commented out this test because the function being tested, getSonarProjectName, is not necessary any more. The project name is easily retrieved from the sonar build report.

File: src/test/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGeneratorTest.java (hunk 5 above)
Comment (dgeorgievski) [reply src/test/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGeneratorTest.java]: Removed for the same reason like above.

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java (hunk 6 above)
Comment (dgeorgievski) [reply src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: Removed

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java (hunk 7 above)
Comment (dgeorgievski) [reply src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: Removed

File: doc/influxdb-v2-integration.md (hunk 2 above)
Comment (asimell) [reply doc/influxdb-v2-integration.md]: Thank you, the documentation is much better with not-so-provocative wording.

If you want to add the token based authentication be my guest, but please create a separate PR for that. I was planning on adding the token authentication "soon" myself, but you're welcome to do it if you have motivation.

Comment (dgeorgievski) [reply doc/influxdb-v2-integration.md]: I'd love to contribute to this project. Wiley is building a CI/CD Dashboard and both Jenkins and this plugin are foundational components for supplying the metrics for the dashboards.  

I’ll start with the Jira board for Jenkins for new ideas, and I’d definitely give a try to adding a Token authentication for InfluxDB v2.

**Comment by dgeorgievski** [comment]
FYI I've explicitly set JUnit `test_count` to `long`. Initially in the doc I defined it mistakenly as a `float`. 

The confusion on my part came from using the defaults for values used by the InfluxDB line protocol, which differ from those used by the InfluxDB Java client.

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java (hunk 7 above)
Comment (dgeorgievski) [reply src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: For some reason the trailing whitespaces escaped my attention . Removed.

**Comment by dgeorgievski** [comment]
Hi @asimell , I believe i've addressed all the requests and questions. Please let me know if there is anything else left. It would be great if we could proceed with the review,

**Comment by asimell** [comment]
Hi @dgeorgievski and sorry for the delay. Yes, everything is addressed for. Unfortunately, when I did an OS upgrade my Docker started to act up and I haven't been able to get my development SonarQube working. I'd really want to test this myself as well before merging. No offence to you or this PR in particular, but I've learned from the past that it saves me a lot of time to test _before_ I do a release. 😅

**Comment by asimell** [comment]
Hi @dgeorgievski

I just tried to run my test jobs and while I no longer get SonarQube errors, no data is sent to `sonarqube_data` as `hasReport()` doesn't recognise the Sonar Scan. It should definitely work for backwards compatibility.

Is `file: ${WORKSPACE}/build/sonar/report-task.txt` something that the Sonar Scanner generates itself or is it something that every user must add themselves to their job?

**Comment by dgeorgievski** [comment]
Hi @asimell,

Considering the complexities of this integration I am glad you are testing the proposed changes as well. 

The location of the report file is defined by [sonar.scanner.metadataFilePath](sonar.scanner.metadataFilePath) setting introduced with version 7.2. 

See these two tickets for the motivation behind this change:
https://jira.sonarsource.com/browse/VSTS-213
https://jira.sonarsource.com/browse/VSTS-190

Prior to v7.2 the `report-task.txt` file has been written at `Agent.BuildDirectory` and it has not been removed after the scan creating issues in rare occasions when the layout of the project or build definition change. 

Initially when I submitted the PR I wasn't aware of this detail and planned to test the changes with various SQ versions, but then I had to take some time off, forced by our HR, before the end of the year and left the testing for this week.  

It looks like you already did that and provided with hints where to look for clues. Thanks again for that.

After reviewing the information available in the SQ tickets it looks to me that it would not make a sense to try to preserve compatibility with pre 7.2 SQ versions due to the simple fact that `report-task.txt` could have multiple instances - one current and other, stale from previous runs.

Secondly, I've tested SQ plugin with Maven and found out that the report is created at `${WORKSPACE}/target/sonar/report-task.txt`

In case of a repo with multiple projects like this [one](https://github.com/dgeorgievski/sonar-scanning-examples/tree/master/sonarqube-scanner-maven), the path could be `${WORKSPACE}/maven-basic/target/sonar/report-task.txt`, which is another strong argument for being able to set the path to the report file explicitly through environment variable like this maybe
```
stage("InfluxDB v2 publisher") {
            environment {
                SONAR_SCANNER_METADATA_FILE_PATH="maven-basic/target/sonar/report-task.txt"
            }
            
            steps {
                withSonarQubeEnv('SonarQube') {
                    influxDbPublisher(selectedTarget: 'influxdb_v2',)
                }
            }
        }
```

The influxdb-plugin could check first this env variable as a hint where to read the report file from. Otherwise it would try common locations for Gradle and Maven projects like
1. Gradle: `${WORKSPACE}/build/sonar/report-task.txt`
2. Maven: `${WORKSPACE}/target/sonar/report-task.txt`
3. Fail if file is not located at any of the above locations.

What do you think? Are you using Maven for testing?

**Comment by asimell** [comment]
The current SonarQube LTS is 8.9, so I don't think we need to support earlier versions.

My current testing jobs (freestyle and pipeline) are configured to run SonarQube scanner directly even though it will run Maven later on.
```
stage('Sonarqube') {
    steps {
        script {
            withSonarQubeEnv('mysonar') {
                scannerHome = tool 'mysonar'
                sh "${scannerHome}/bin/sonar-scanner"
            }
        }
    }
}
// ...
stage('Report') {
    steps {
        script {
            withSonarQubeEnv('mysonar') {
                influxDbPublisher(selectedTarget: 'Target 1')
            }
        }
    }
}
```

However, I must say my job configurations date back to SonarQube 5.x, so my configurations might not work out of the box. Still, I'd find it convenient if the plugin could find the SonarQube reports without having to configure the jobs separately.

`sonar.scanner.metadataFilePath` is new to me. I will try to modify my `sonar-project.properties` file to have that and see if I find the results.

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java [hunk 8]
```diff
@@ -49,29 +48,34 @@
     private static final short DEFAULT_MAX_RETRY_COUNT = 10;
     private static final int DEFAULT_RETRY_SLEEP = 5000;
 
-    private static final String URL_PATTERN_IN_LOGS = ".*" + Pattern.quote("ANALYSIS SUCCESSFUL, you can browse ")
-            + "(.*)";
-    private static final String TASK_ID_PATTERN_IN_LOGS = ".*" + Pattern.quote("More about the report processing at ")
-            + "(.*)";
+    // patterns used for data extraction from ${WORKSPACE}/${SONAR_BUILD_REPORT} file
+    private static final String SONAR_BUILD_REPORT = "/build/sonar/report-task.txt";
```
Comment (asimell) [inline src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: This looks dangerous.

SonarQube's [official documentation](https://docs.sonarqube.org/latest/analysis/analysis-parameters/) says the default value for `sonar.scanner.metadataFilePath` is the value for `sonar.working.directory`, which in turn has a default value of `.scannerwork`. I think the default value should correspond to that, but I think it could be overwritable with an environment variable.

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java [hunk 9]
```diff
@@ -80,21 +84,30 @@ public SonarQubePointGenerator(Run<?, ?> build, TaskListener listener,
...
 
+    /**
+     * @return true, if environment variable LOG_SONAR_QUBE_RESULTS is set to true and SQ Reports exist
+     */
     public boolean hasReport() {
-        try {
-
-            String[] result = getSonarProjectURLFromBuildLogs(build);
+        String logMessage = "[InfluxDB Plugin] SonarQube hasReport";
+        listener.getLogger().println(logMessage);
```
Comment (asimell) [inline src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: This seems like an irrelevant log message outside development debugging. If SonarQube report is found it will log `SonarQube data found. Writing to InfluxDB...` anyway in `InfluxDbPublicationService.java`.

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java (hunk 8 above)
Comment (asimell) [inline src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: `.scannerwork` is the default directory for SonarQube analysis
```suggestion
    private static final String SONAR_BUILD_REPORT = "/.sonarscanner/report-task.txt";
```

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java [hunk 10]
```diff
@@ -253,45 +248,55 @@ protected String getResult(String url) throws IOException {
...
-    private String[] getSonarProjectURLFromBuildLogs(Run<?, ?> build) throws IOException {
+    private String[] getSonarProjectFromBuildReport() throws IOException {    
+        String projName = null;
         String url = null;
-        String taskid = null;
-
-        try (BufferedReader br = new BufferedReader(build.getLogReader())) {
+        String taskId = null;
+        String taskUrl = null;
+        String file = env.get("WORKSPACE", "") + SONAR_BUILD_REPORT;
```
Comment (asimell) [inline src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: Add support for `SONAR_SCANNER_METADATA_FILE_PATH`. This should then be documented in SonarQube integration guidelines.
```suggestion
        String default_file = env.get("WORKSPACE", "") + SONAR_BUILD_REPORT;
        String file = env.get("SONAR_SCANNER_METADATA_FILE_PATH", default_file);
```

**Comment by dgeorgievski** [comment]
Hi @asimell, I just wanted to acknowledge your feedback. Surprisingly, I am going though a busy period at work and haven't been able to find time for this work.

**Comment by dgeorgievski** [comment]
Hi @asimell , I am back to work on this issue. I am able now to test various SQ versions, but I agree let's target the  current SQ LTS v8.9 at a minimum for this solution. 

I agree with all your remarks and will try to address them next.

**Comment by dgeorgievski** [comment]
Hi @asimell,

After reviewing SQ documentation again, I've decided to implement a function to find the report file created by the scanner. Its location varies depending on the build system. 

The only assumption made is the report file name - `report-task.txt`. If for whatever reason the file name changes one could use the `SONARQUBE_BUILD_REPORT_NAME` to specify the new name of the report file. Please see the integration doc for details.

I was able to successfully test the search functionality with Maven and Gradle projects using SQ 8.6. I believe I'll be able test the integration later with .Net projects as well.

I've also added unit tests to verify the search functionality with a temp file system, which should make future improvements more predictable and reliable. Initially I wanted to add the new tests as part of a nested class, but that comes in JUnit 5 and the project is using version 4. It would be nice if JUnit is upgraded to v5 in the future.

Little off topic. It would have been also nice if plugins in different stages could report their state using some kind of messaging services or maybe use events to execute stages in a pipeline like https://keptn.sh/. Messaging would help decouple plugins and  eliminate the need to know how a plugin manages files internally. Just curious, still learning about Jenkins internals, are there any plans for improving the architecture of pipelines?

**Comment by dgeorgievski** [comment]
Looks like the Jenkins CI failure was caused by a unavailable Windows EC2 instance
```
[2022-01-16T00:31:09.248Z] EC2 (aws-us-east-2) - Windows 2019 (i-0c72ec9b0f133ee35) was marked offline: Connection was broken: java.io.EOFException

[2022-01-16T00:31:09.248Z] 	at java.base/java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2872)

[2022-01-16T00:31:09.248Z] 	at java.base/java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3367)

[2022-01-16T00:31:09.248Z] 	at java.base/java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:936)

[2022-01-16T00:31:09.248Z] 	at java.base/java.io.ObjectInputStream.<init>(ObjectInputStream.java:379)

[2022-01-16T00:31:09.248Z] 	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:49)

[2022-01-16T00:31:09.248Z] 	at hudson.remoting.Command.readFrom(Command.java:142)

[2022-01-16T00:31:09.248Z] 	at hudson.remoting.Command.readFrom(Command.java:128)

[2022-01-16T00:31:09.248Z] 	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)

[2022-01-16T00:31:09.248Z] 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:61)

[2022-01-16T00:31:09.248Z] Caused: java.io.IOException: Unexpected termination of the channel

[2022-01-16T00:31:09.248Z] 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:75)
```

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java (hunk 8 above)
Comment (dgeorgievski) [reply src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: Not relevant with the new changes.

File: src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java (hunk 10 above)
Comment (dgeorgievski) [reply src/main/java/jenkinsci/plugins/influxdb/generators/SonarQubePointGenerator.java]: Not relevant with the new changes.

File: doc/influxdb_v2_integration.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
+# InfluxDB v2 Integration
+
+The current plugin version uses InfluxDB v1.x authentication semantics (user name and password, and a database), while v2 is using API token, bucket, and organization to regulate the access to the data. 
+
+The current Jenkins InfluxDB Target configuration requires the following: 
+- A Jenkins Credential with username and password, instead of a Secret Text that would better store the ImfluxDB v2 token. 
+- InfluxDB v2 settings: bucket, org, and a secret. If the org is not specified the plugin creates an InfluxDB v1 client. 
+- Performs HTTP Basic authentication with the provided username and password credentials. The password is significant but the username is not which the Basic authentication scheme used by InfluxDB v2 client, which results in an authentication failure. 
+
+The solution is to use InfluxDB Telegraf http listener that serves as a proxy between the influxdb_plugin and InfluxDB v2 instance. The listener can authenticate the influxdb_plugin client using only the password with a Basic schema, and stream the metrics to InfluxDB v2 with the help of influxdb_v2 output plugin, that in turn could perform filtering on the measurements and tags.
```
Comment (asimell) [inline doc/influxdb_v2_integration.md]: Since #125 fixed InfluxDB 2.x authentication with username+password, this is not entirely true anymore. You _can_ authenticate to InfluxDB 2.x with username+password even though the standard way of authenticating is by token (which is still under work).

Comment (dgeorgievski) [reply doc/influxdb_v2_integration.md]: Agreed.  I've just tested successfully  username+password authentication.

Initially, this was my stumbling block that was quickly eclipsed by SQ integration issues. Now that i am more familiar with how the plugin works, it is easier to better articulate the requirements for both integrations.

**Review by asimell — CHANGES_REQUESTED** [review]
Since https://github.com/jenkinsci/influxdb-plugin/pull/126 adds support for token based authentication, the `influxdb_v2_integration.md` file is obsolete. Of course one could still use Telegraf, but it's not related to SonarQube fixes and should be removed from this PR. You can open another PR with this documentation file if you want.

**Comment by dgeorgievski** [comment]
> Since #126 adds support for token based authentication, the `influxdb_v2_integration.md` file is obsolete. Of course one could still use Telegraf, but it's not related to SonarQube fixes and should be removed from this PR. You can open another PR with this documentation file if you want.

Sounds good with me. Removed`influxdb_v2_integration.md` from this PR.

**Review by asimell — APPROVED** [review]
LGTM

**Comment by kumbasar** [comment]
Hello @asimell  & @dgeorgievski ,

Unfortunately, this merge request has broken our Jenkins/SonarQube - InfluxDB integration :-/

I have added some nasty debug logs and found out that the plugin is searching the `report-task.txt` in the Jenkins server (aka Built-In Node), not in the Jenkins builders (nodes) where the pipeline executed:

image

Therefore, it cannot find the report and skips the SonarQube part. In my case, Jenkins executes the job in a docker container. Anyway. 

Do you have any idea how to fix this issue?
